mvc/src/main/java/nextstep/mvc/controller/tobe/AnnotationHandlerMapping.java

```diff
@@ -1,30 +1,93 @@
 package nextstep.mvc.controller.tobe;
 
 import jakarta.servlet.http.HttpServletRequest;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
 import nextstep.mvc.HandlerMapping;
+import nextstep.mvc.exception.NoDefaultConstructorException;
+import nextstep.web.annotation.Controller;
+import nextstep.web.annotation.RequestMapping;
+import nextstep.web.support.RequestMethod;
+import org.reflections.Reflections;
+import org.reflections.util.ClasspathHelper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.HashMap;
-import java.util.Map;
-
 public class AnnotationHandlerMapping implements HandlerMapping {
 
     private static final Logger log = LoggerFactory.getLogger(AnnotationHandlerMapping.class);
 
-    private final Object[] basePackage;
+    private final Object[] basePackages;
     private final Map<HandlerKey, HandlerExecution> handlerExecutions;
 
-    public AnnotationHandlerMapping(final Object... basePackage) {
-        this.basePackage = basePackage;
+    public AnnotationHandlerMapping(final Object... basePackages) {
+        this.basePackages = basePackages;
         this.handlerExecutions = new HashMap<>();
     }
 
     public void initialize() {
         log.info("Initialized AnnotationHandlerMapping!");
+
+        for (final Object basePackage : basePackages) {
+            final Collection<URL> packagePath = ClasspathHelper.forPackage(basePackage.toString());
+            final Reflections reflections = new Reflections(packagePath);
+            final Set<Class<?>> controllers = reflections.getTypesAnnotatedWith(Controller.class);
+
+            mapControllerForHandlerExecution(controllers);
+        }
```

Comment:
```suggestion
        final Reflections reflections1 = new Reflections(basePackages);
        final Set<Class<?>> typesAnnotatedWith = reflections1.getTypesAnnotatedWith(Controller.class);
        mapControllerForHandlerExecution(typesAnnotatedWith);
        }
```

a : 저는 이런식으로 여러 Object를 한번에 받도록 처리했어요! 이런 방법도 있다~ 정도로 봐주세용

Q : `new Reflections()` 로 생성하면 어떻게 될까요? 정상적으로 동작이 되지 않을까요?? 동작이 된다면 내부적으로 어떻게 처리하는 걸까요?

Reply:
Reflections 생성자에서 Object 타입을 가변인자로 받고 있어서 반복문을 사용할 필요 없이 `basePackages`를 바로 넘겨줄 수 있었네요ㅎㅎ


> Reflections.java

<img width="401" alt="image" src="https://user-images.githubusercontent.com/28749734/191240766-85ad7b55-40c1-4e74-b033-8825a208a1b1.png">

생성자에서 `ConfigurationBuilder`로 파라미터를 하나씩 처리해주고

> ConfigurationBuilder.java

빌더에서는 인자에 대한 패키지를 각각 찾아서

<img width="958" alt="image" src="https://user-images.githubusercontent.com/28749734/191241082-7709d3d7-a1aa-4c96-93b2-5ecc765831ee.png">

제가 사용하던 `ClasspathHelper.forPackage()`가 여기서 사용되는군요!

<img width="659" alt="image" src="https://user-images.githubusercontent.com/28749734/191241138-02161970-c860-4610-8435-486f737cb241.png">

로마의 리뷰 덕분에 Reflections 생성자의 동작 원리도 꼼꼼하게 살펴볼수 있었어요👍

Reply:
꼼꼼하게 학습하셨군요.. 고생하셨습니다. 🧚‍♀️